Sources/NIOPosix/PooledRecvBufferAllocator.swift

```diff
@@ -0,0 +1,192 @@
+                    } else {
+                        self.buffers.formIndex(after: &index)
+                    }
+                }
+
+                // Sweep up from the start to the last used index.
+                index = self.buffers.startIndex
+                while index != startIndex {
+                    if let result = try self.buffers[index].modifyIfUniquelyOwned(minimumCapacity: nextBufferSize, body) {
+                        self.lastUsedIndex = index
+                        return (self.buffers[index], result)
+                    } else {
+                        self.buffers.formIndex(after: &index)
+                    }
+                }
+            }
+        } else if self.buffer != nil, !self.mayGrow {
+            // No hint about the buffer size (so pooling is not being used) and the allocator
+            // indicated that the next buffer will not grow in size so reuse the existing stored
+            // buffer.
+            self.buffer!.clear()
+            let result = try body(&self.buffer!)
+            return (self.buffer!, result)
+        }
+
+        // No available buffers or the allocator does not offer up buffer sizes; directly
+        // allocate a new one.
+        var newBuffer = self.recvAllocator.buffer(allocator: allocator)
+
+        if let buffer = self.buffer {
+            assert(self.buffers.isEmpty)
+            // Have a stored buffer, either:
+            // 1. We have capacity to add more and use `buffers` for storage, or
+            // 2. Our capacity is 1; we can't use `buffers` for storage.
+            if self.capacity > 1 {
+                self.buffer = nil
+                self.buffers.reserveCapacity(self.capacity)
+                self.buffers.append(buffer)
+                self.buffers.append(newBuffer)
+                self.lastUsedIndex = self.buffers.index(before: self.buffers.endIndex)
+                return try self.modifyLast(body)
+            } else {
+                let result = try body(&newBuffer)
+                return (newBuffer, result)
+            }
+        } else {
+            // There's no stored buffer which could be due to:
+            // 1. this is the buffer we allocate (i.e. buffers is empty), or
+            // 2. we've already switched to using buffers for storage and it's not yet full, or
+            // 3. we've already switched to using buffers for storage and it's full.
+            if self.buffers.isEmpty {
+                self.buffer = newBuffer
+                let result = try body(&self.buffer!)
+                return (self.buffer!, result)
+            } else if self.buffers.count < self.capacity {
+                self.buffers.append(newBuffer)
+                self.lastUsedIndex = self.buffers.index(before: self.buffers.endIndex)
+                return try self.modifyLast(body)
+            } else {
+                let result = try body(&newBuffer)
+                return (newBuffer, result)
+            }
+        }
+    }
```

Comment:
I think this method is beginning to get a bit unwieldy. Can we refactor it into smaller helper functions that each handle one case?